orchestra/actions/action.py

```diff
@@ -17,12 +17,12 @@ def __init__(self, name, script, config):
         self._explicit_dependencies: Set[Action] = set()
         self._script = script
 
-    def run(self, cmdline_args, *args, **kwargs):
+    def run(self, pretend=False, **kwargs):
         logger.info(f"Executing {self}")
-        if not cmdline_args.pretend:
-            self._run(cmdline_args, *args, **kwargs)
+        if not pretend:
+            self._run(**kwargs)
 
-    def _run(self, cmdline_args):
+    def _run(self, **kwargs):
         """Executes the action"""
         self._run_user_script(self.script)
```

Comment:
I'm confused, what are the `**kwargs` for?

Reply:
It's so that subclasses can take kwargs without the type checker giving warnings.
InstallAction has `def _run(self, set_manually_installed=False)`, because Executor knows whether the component should be set as "manually installed", which is defined as "the user has requested to install this component explicitly"

Reply:
If they have such a different signature, are you sure they are overloads?
Shouldn't they have a different name maybe?

Reply:
I could change this
```python
class Executor:
    ...
    def _run_action(self, action: Action):
        self._running_actions.append(action)
        self._current_remaining -= 1
        kwargs = {}
        if isinstance(action, InstallAction):
            kwargs["set_manually_installed"] = action in self.actions

        try:
            return action.run(pretend=self.pretend, **kwargs)
        finally:
            self._running_actions.remove(action)
```
into this
```python
class Executor:
    ...
    def _run_action(self, action: Action):
        self._running_actions.append(action)
        self._current_remaining -= 1
        if isinstance(action, InstallAction):
            action.set_manually_installed = action in self.actions

        try:
            return action.run(pretend=self.pretend)
        finally:
            self._running_actions.remove(action)
```

What do you think?

Reply:
Mmmh, I think all `Action.run` methods should just take an extra argument such as `explicitly_requested` and then it's up to them to decide what to do with that information.
`Executor` should not do special handling for `InstallAction`.